lib/rules/array-bracket-spacing.js

```diff
@@ -123,17 +123,19 @@ module.exports = function(context) {
         var first = context.getFirstToken(node),
             second = context.getFirstToken(node, 1),
             penultimate = context.getLastToken(node, 1),
-            last = context.getLastToken(node);
+            last = context.getLastToken(node),
+            firstElement = node.elements[0],
+            lastElement = node.elements[node.elements.length - 1];
 
         var openingBracketMustBeSpaced =
-            options.objectsInArraysException && second.value === "{" ||
-            options.arraysInArraysException && second.value === "[" ||
+            options.objectsInArraysException && (firstElement.type === "ObjectExpression" || firstElement.type === "ObjectPattern") ||
```

Comment:
Do you mean to extract a check for each of the types (object and array) or a `mustElementBeSpaced(element)` function to extract the whole check?